lib/utils/change-object.js

```diff
@@ -1,15 +1,34 @@
 var assign = require('lodash/assign')
+var validate = require('../validate')
 
 /**
   * change object either by passing changed properties
   * as an object, or by passing a change function that
   * manipulates the passed object directly
   **/
-module.exports = function changeObject (change, object) {
+module.exports = function changeObject (state, change, object, validateChange) {
```

Comment:
I wonder if we can avoid this extra argument. It looks like we only use it for mark-as-deleted. But why shouldn’t we validate when deleting?

Reply:
That is something I'd like to discuss too. Delete operation's purpose is more of a data clean-up utility. Why would we want to validate whether the doc passed can be marked for deletion? Are there any scenarios which come to your mind?
Delete should be unconditional, isn't it?

Reply:
Technically, a delete in CouchDB/Hoodie land is just an update. the update method also accepts an optional change so you can for example add context to the deletion. For example, you could check for consistency, like not allowing to delete a todo list while there are still todos assigned to it?

Reply:
But that can be caught during the update flow right? `changeObject` will be referenced twice, once for adding the marked-for-deletion entry and the other one for `update` which will have validation on it.
From what you just told me, deletion is infact an update scenario.

Reply:
So the major doubt here is, should we consider remove/update seperate operations or similar?
I'm also concerned about the how a single validation function will scale for all these different operations 
...

Reply:
> I'm also concerned about the how a single validation function will scale for all these different operations

it’s a problem when it’s a problem :) in most cases I think the validate function will be super simple and work add/update/remove. I’d say keep it simple, then make it complex as needed

Reply:
What path should we take then @gr2m ?
* Add validate to remove , in which case, the doc will go through two validations, one for remove and the other for update
* Keep it like this at the moment

Reply:
🤔 

Maybe we would put in the validation logic into `change-object.js` as its called for both update & remove, but that method currently synchronous, so we’d need to change that.

But it _feels_ like the right place to run validation. What you think?

Reply:
I have put in validation flow in `change-object.js` 
That's why validation will trigger twice for remove  operations, if I remove the `validateChange` hack/logic.

Reply:
I’ll give this another review over the weekend, sorry to make you wait that long